write-fonts/src/tables/glyf.rs

```diff
@@ -0,0 +1,298 @@
+pub trait OtPoint {
+    fn get(self) -> (i16, i16);
+}
+
+impl OtPoint for kurbo::Point {
+    fn get(self) -> (i16, i16) {
+        (ot_round(self.x as f32), ot_round(self.y as f32))
+    }
+}
+
+impl OtPoint for (i16, i16) {
+    fn get(self) -> (i16, i16) {
+        self
+    }
+}
+
+// adapted from simon:
+// https://github.com/simoncozens/rust-font-tools/blob/105436d3a617ddbebd25f790b041ff506bd90d44/otmath/src/lib.rs#L17
+fn ot_round(val: f32) -> i16 {
+    (val + 0.5).floor() as i16
+}
+
+impl Contour {
+    /// Create a new contour begining at the provided point
+    pub fn new(pt: impl OtPoint) -> Self {
+        let (x, y) = pt.get();
+        Self(vec![CurvePoint::on_curve(x, y)])
+    }
+
+    /// The total number of points in this contour
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// `true` if this contour is empty
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Add a line segment
+    pub fn line_to(&mut self, pt: impl OtPoint) {
+        let (x, y) = pt.get();
+        self.0.push(CurvePoint::on_curve(x, y));
+    }
+
+    /// Add a quadratic curve segment
+    pub fn curve_to(&mut self, p0: impl OtPoint, p1: impl OtPoint) {
+        let (x0, y0) = p0.get();
+        let (x1, y1) = p1.get();
+        self.0.push(CurvePoint::off_curve(x0, y0));
+        self.0.push(CurvePoint::on_curve(x1, y1));
+    }
+}
+
+impl SimpleGlyf {
+    /// Attempt to create a simple glyph from a kurbo `BezPath`
+    ///
+    /// The path may contain only line and quadratic bezier segments. The caller
+    /// is responsible for converting any cubic segments to quadratics before
+    /// calling.
+    ///
+    /// Returns an error if the input path is malformed; that is, if it is empty,
+    /// contains cubic segments, or does not begin with a 'move' instruction.
+    //TODO: figure out a more general API? maybe a nested builder thing, where you
+    //build contours, and from those contours build a glyph? idk?
+    pub fn from_kurbo(path: &BezPath) -> Result<Self, BadKurbo> {
+        let mut contours = Vec::new();
+        let mut current = None;
+
+        for el in path.elements() {
+            match el {
+                kurbo::PathEl::MoveTo(pt) => {
+                    if let Some(prev) = current.take() {
+                        contours.push(prev);
+                    }
+                    current = Some(Contour::new(*pt));
+                }
+                kurbo::PathEl::LineTo(pt) => {
+                    current.as_mut().ok_or(BadKurbo::MissingMove)?.line_to(*pt)
+                }
+                kurbo::PathEl::QuadTo(p0, p1) => current
+                    .as_mut()
+                    .ok_or(BadKurbo::MissingMove)?
+                    .curve_to(*p0, *p1),
+                kurbo::PathEl::CurveTo(_, _, _) => return Err(BadKurbo::HasCubic),
+                // I think we can just ignore this, and remove duplicate points
```

Comment:
All contours are implicitly closed so this seems reasonable.

Reply:
Would that duplicate point removal ever make something meant to be interpolation friendly not be?

Reply:
It shouldn't; it is addressing the fact that in ot all contours have the same start/end point, but in kurbo a bezpath is composed of segments, and if you want a 'closed' contour you generally finish with a segment that has an end point equal to your starting point.

Reply:
Rod actually raises a valid point here. I haven't checked this code too deeply, but I recall in fonttools we had a similar issue in the converter pen that translates from point-wise to segment-wise representation of a path, and we needed to special case the situation when the last segment of a closed contour is a "line" and its point overlaps with the contour's "move" point: https://github.com/fonttools/fonttools/pull/1720/

In the sources, which contain list of points, as well as in the output glyf table, which also contains only points, it's clear and non ambiguous whether there's or not a duplicate point at the end/start of the contour; when translating points to segment and back to points it's important to not lose this duplicate points, because they may be intentional to support clever interpolation of variable contours (e.g. where a line segment "disappears" in some masters)

Reply:
To give a simple example, consider a box that becomes a triangle by moving two corners to a point in between. Those points - if at the end - are now duplicate for the triangle case but not the square. If you drop them you'd make paths that used to interpolate not do so anymore.